flink-clients/src/main/java/org/apache/flink/client/program/StreamContextEnvironment.java

```diff
@@ -169,12 +233,53 @@ public static void setAsContext(
                             mergedConfiguration,
                             userCodeClassLoader,
                             enforceSingleJobExecution,
-                            suppressSysout);
+                            suppressSysout,
+                            allowConfigurations,
+                            errors);
                 };
         initializeContextEnvironment(factory);
     }
 
     public static void unsetAsContext() {
         resetContextEnvironment();
     }
+
+    private List<String> collectNotAllowedConfigurations() {
+        final List<String> errors = new ArrayList<>();
+        if (allowConfigurations) {
+            return errors;
+        }
+        final MapDifference<String, String> diff =
+                Maps.difference(originalConfiguration.toMap(), configuration.toMap());
+        diff.entriesOnlyOnRight()
+                .forEach(
+                        (k, v) ->
+                                errors.add(
+                                        ConfigurationNotAllowedMessage.ofConfigurationKeyAndValue(
+                                                k, v)));
+
+        if (!Arrays.equals(originalCheckpointConfigSerialized, serializeConfig(checkpointCfg))) {
+            errors.add(
+                    ConfigurationNotAllowedMessage.ofConfigurationObject(
+                            originalCheckpointConfigSerialized.getClass().getName()));
```

Comment:
nit: Do we need to print the fully qualified class name or would the simple name be enough? Maybe it would make sense to pass the class object here and let `ConfigurationNotAllowedMessage` take care of the formatting.
```suggestion
                    ConfigurationNotAllowedMessage.ofConfigurationObject(
                            originalCheckpointConfigSerialized.getClass());
```